File: CausalInference/Udemy_Econometrics_and_Statistics_for_Business_in_R_n_Python/3. Granger Causality/Granger Causality.R
#pull the data
#install.packages("lmtest")
library(lmtest)
dataset <- as.data.frame(ChickEgg)

#tranform data into time series
eggs <- ts(dataset$egg)
chickens <- ts(dataset$chicken)

#plotting time series
#install.packages("ggplot2")
library(ggplot2)
plot.ts(eggs)
plot.ts(chickens)

#Stationarity check
#install.packages("forecast")
library(forecast)
ndiffs(eggs, alpha = 0.05, test = c("kpss"))
ndiffs(chickens, alpha = 0.05, test = c("kpss"))

#make data stationary
deggs <- diff(eggs)
dchickens <- diff(chickens)

#plotting time series
plot.ts(deggs)
plot.ts(dchickens)

#Do granger causality
grangertest(deggs, dchickens, 4)
grangertest(dchickens, deggs, 4)

#start looping process
lags <- 1:15
results <- matrix(ncol = 2, nrow = length(lags))

#start looping
for (i in 1:length(lags)){
  test1 <- grangertest(deggs, dchickens, i)
  result1 <- test1$`Pr(>F)`[2]
  test2 <- grangertest(dchickens, deggs, i)
  result2 <- test2$`Pr(>F)`[2]  
  
  results[i,1] <- result1
  results[i,2] <- result2
  
}

File: CausalInference/Udemy_Econometrics_and_Statistics_for_Business_in_R_n_Python/3. Granger Causality/Granger Causality.py
#importing the libraries
import pandas as pd
from matplotlib import pyplot

#import the dataset
dataset = pd.read_csv("chicken_egg.csv")

#isolating eggs and chickens
eggs = dataset.loc[:, 'egg']
chickens = dataset.loc[:, 'chicken']

#visualize eggs
eggs.plot()
pyplot.show()

#visualize chickens
chickens.plot()
pyplot.show()

#check stationarity for eggs
from statsmodels.tsa.stattools import adfuller
stationarity_eggs = adfuller(eggs)
stationarity_eggs
print('Augment Dickez Fuller p-value: %F' % stationarity_eggs[1])

#check stationarity for chickens
stationarity_chickens = adfuller(chickens)
stationarity_chickens
print('Augment Dickez Fuller p-value: %F' % stationarity_chickens[1])

#making data stationary
deggs = eggs - eggs.shift(1)
deggs = deggs.dropna()
dchickens = chickens - chickens.shift(1)
dchickens = dchickens.dropna()

#check stationarity 
stationarity_deggs = adfuller(deggs)
print('p-value: %F' % stationarity_deggs[1])
stationarity_dchickens = adfuller(dchickens)
print('p-value: %F' % stationarity_dchickens[1])

#transforming series into dataframes
deggs = pd.DataFrame(deggs)
dchickens = pd.DataFrame(dchickens)

#creating two dataframes
dataset_chicken = pd.concat([dchickens, deggs], axis = 1)
dataset_eggs = pd.concat([deggs, dchickens], axis = 1)

#do granger causality test
from statsmodels.tsa.stattools import grangercausalitytests
grangertest_chickens = grangercausalitytests(dataset_chicken, 15)
grangertest_eggs = grangercausalitytests(dataset_eggs, 15)































































